test/utils/azure/azure_helpers.go

```diff
@@ -245,6 +247,123 @@ func (az *Client) GetVirtualNetworkSubnet(ctx context.Context, groupName, vnetNa
 	return az.subnetsClient.Get(ctx, groupName, vnetName, subnetName, "")
 }
 
+func (az *Client) GetNodeVM(ctx context.Context, groupName, nodeName string) (vm compute.VirtualMachine, err error) {
+	vms, err := az.ListVMs(ctx, groupName)
+	if err != nil || len(vms) == 0 {
+		return vm, fmt.Errorf("getting the vm list failed or no vms exist in the resource group: %v", err)
+	}
+	for _, vm := range vms {
+		if nodeName == *vm.VirtualMachineProperties.OsProfile.ComputerName {
+			return az.vmClient.Get(ctx, groupName, *vm.Name, compute.InstanceView)
+		}
+	}
+	return vm, fmt.Errorf("no matching vm instance was found for the node %s: %v", nodeName, err)
+}
+
+func (az *Client) GetNodeVMSSNameAndVM(ctx context.Context, groupName, nodeName string) (vmssvm compute.VirtualMachineScaleSetVM, vmssName string, err error) {
+	vmsss, err := az.ListVMSSs(ctx, groupName)
+	if err != nil || len(vmsss) == 0 {
+		return vmssvm, vmssName, fmt.Errorf("getting the vmss list failed or no vmss exist in the resource group: %v", err)
+	}
+	for _, vmss := range vmsss {
+		vmssvms, err := az.vmssvmClient.List(ctx, groupName, *vmss.Name, "", string(compute.InstanceView), string(compute.InstanceView))
+		if err != nil {
+			return vmssvm, vmssName, fmt.Errorf("getting the vm list from vmss failed: %v", err)
+		}
+		for _, vm := range vmssvms.Values() {
+			vmssvm, err := az.vmssvmClient.Get(ctx, groupName, *vmss.Name, *vm.InstanceID, compute.InstanceView)
+			if err != nil {
+				return vmssvm, vmssName, fmt.Errorf("getting the vm instance from vmss failed: %v", err)
+			}
+			if nodeName == *vmssvm.OsProfile.ComputerName {
+				return vmssvm, *vmss.Name, nil
+			}
+		}
+	}
+	return vmssvm, vmssName, fmt.Errorf("no matching vmss vm instance was found for the node %s: %v", nodeName, err)
+}
+
+func (az *Client) ListVMs(ctx context.Context, groupName string) ([]compute.VirtualMachine, error) {
+	vms, err := az.vmClient.List(ctx, groupName)
+	if err != nil {
+		return nil, fmt.Errorf("getting the vm list failed: %v", err)
+	}
+	return vms.Values(), nil
+}
+
+func (az *Client) ListVMSSs(ctx context.Context, groupName string) ([]compute.VirtualMachineScaleSet, error) {
+	vmss, err := az.vmssClient.List(ctx, groupName)
+	if err != nil {
+		return nil, fmt.Errorf("getting the vmss list failed: %v", err)
+	}
+	return vmss.Values(), nil
+}
+
+func (az *Client) AttachDiskToVMSSVMInstance(ctx context.Context, groupName, diskName, volumeID, vmssName string, vmssvm compute.VirtualMachineScaleSetVM) error {
+	dataDisk := compute.DataDisk{
+		Lun:          to.Int32Ptr(0),
```

Comment:
this would lead to error easily since it assumes there is no disk using LUN 0, what about:
1. cordon node#1
2. `azurediskDriver.ControllerPublishVolume` to attach disk to node#0 (since node#0 is cordoned)
3. uncordon node#1, cordon#0
4. create pod with that disk

By this way, you don't even need to write attach disk API

Reply:
This is the same thing I was wondering and discussed with you briefly over Teams. Using ControllerPublishVolume certainly seems like a better and easier option, but cardoning nodes will not be sufficient from what I understand. As I noted before, ControllerPublishVolume [requires NodeId](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/pkg/azuredisk/controllerserver.go#L379) as a parameter and this Id is not trivial to obtain. Let me know if I am missing anything.

Reply:
since you already use k8s api `ListNodes()`, the nodeID should be in the returned values, for vmss, its format is like `k8s-agentpool-37051515-vmss000001`, same as what you get with `kubectl get no`

Reply:
That only works for VMSets, for VMSS like I mentioned in our discussion, it returns "instance not found" error. The problem, from what I understand, is in getNodeVMSet function which misclassifies VMSS as VMSet and thus fails to find the node. If what you describe is the expected behavior, maybe I should first work on fixing the behavior of getNodeVMSet?

Reply:
Just to clarify few things and give more context, initially when the ControllerPublishVolume failed with "instance not found", I attributed it to the fact that I am passing the wrong node id and that it expect the id from nodeClient.NodeGetInfo. Now, that I succeeded running it with a different setup, VMSet instead of VMSS(my initial setup), I realized that as you mentioned, passing the node name works! So, when investigating why it doesn't work with VMSS, I saw that in my setup VMType is not specified in the config, so it defaults to VMType = "standard" instead of "vmss". When this value is set correctly to "vmss", everything works as expected. Now what I am trying to understand is, is this unique to my setup or is this a bug that needs to be addressed first? Do we always expect VMType to be provided from azure?

Reply:
AKS does not allow VMSet and VMSS in one cluster, if it's VMSet cluster, then `VMType = "standard"` otherwise it's `VMType = "VMSS"`, will that work?

Reply:
The thing is that it is not in the same cluster. It is exclusively vmss cluster but as it doesn't find the type defined in the config(`$KUBECONFIG`), it defaults to `VMType = "standard"`, causing the problem described.

Reply:
related to this long term issue: https://github.com/kubernetes-sigs/azuredisk-csi-driver/issues/576, while for this test case, could have a simple workaround